File: second_year/plf/exam_prep/course_material/Lecture_06/example_2.pl
month("January").
month("February").
month("March").
month("April").
month("May").
month("June").
month("July").
month("August").
month("September").
month("October").
month("November").
month("December").

month("January",1).
month("February",2).
month("March",3).
month("April",4).
month("May",5).
month("June",6).
month("July",7).
month("August",8).
month("September",9).
month("October",10).
month("November",11).
month("December",12).

isMember([H|_], H).
isMember([_|T], H) :- isMember(T, H).

date(D,M,Y):- D > 0, D < 29, month(M), integer(Y),!.
date(D,M,Y):- D > 0, D < 31, month(M,N), isMember([4,6,9,11],N), integer(Y),!.
date(D,M,Y):- D > 0, D < 32, month(M,N), isMember([1,3,5,7,8,10,12],N), integer(Y),!.
date(29,"February",Y):- Y mod 400 =:= 0, !.
date(29,"February",Y):- Y mod 4 =:= 0, Y mod 100 =\= 0.

less(date(_,_,Y1), date(_,_,Y2)):- Y1 < Y2, !.
less(date(_,M1,Y), date(_,M2,Y)):- month(M1,N1), month(M2,N2), N1 < N2, !.
less(date(D1,M,Y), date(D2,M,Y)):- D1 < D2.

successor(date(D,M,Y), date(DD,M,Y)):- date(D,M,Y), DD is D+1, date(DD,M,Y), !.
successor(date(D,M,Y), date(1,MM,Y)):- date(D,M,Y), month(M,N), NN is N + 1, month(MM,NN), date(1,MM,Y), !.
successor(date(D,M,Y), date(1,"January",YY)):- date(D,M,Y), YY is Y+1.

predecessor(date(D,M,Y), date(DD,M,Y)):- date(D,M,Y), DD is D-1, date(DD,M,Y), !.
predecessor(date(D,M,Y), date(31,MM,Y)):- date(D,M,Y), month(M,N), NN is N - 1, month(MM,NN), date(31,MM,Y), !.
predecessor(date(D,M,Y), date(30,MM,Y)):- date(D,M,Y), month(M,N), NN is N - 1, month(MM,NN), date(30,MM,Y), !.
predecessor(date(D,M,Y), date(29,MM,Y)):- date(D,M,Y), month(M,N), NN is N - 1, month(MM,NN), date(29,MM,Y), !.
predecessor(date(D,M,Y), date(28,MM,Y)):- date(D,M,Y), month(M,N), NN is N - 1, month(MM,NN), date(28,MM,Y), !.
predecessor(date(D,M,Y), date(31,"December",YY)):- date(D,M,Y), YY is Y-1.

%diffDays(Date1, Date2, Diff):- py_call(example_2:daysDiff(Date1, Date2), Diff).
%diffDays(Date1, Date1, 0).
%diffDays(Date1, Date2, Diff):- less(Date1, Date2), successor(Date1, D1), diffDays(D1, Date2, Diff1), Diff is Diff1 + 1.
%diffDays(Date1, Date2, Diff):- less(Date2, Date1), predecessor(Date1, D1), diffDays(D1, Date2, Diff1), Diff is Diff1 - 1.

daysFromDate(Date1, Diff, Date2):- diffDays(Date1, Date2, Diff).

weekday("Monday"):-!.
weekday("Tuesday"):-!.
weekday("Wednesday"):-!.
weekday("Thursday"):-!.
weekday("Friday"):-!.
weekday("Saturday"):-!.
weekday("Sunday"):-!.

weekday("Monday",1):-!.
weekday("Tuesday",2):-!.
weekday("Wednesday",3):-!.
weekday("Thursday",4):-!.
weekday("Friday",5):-!.
weekday("Saturday",6):-!.
weekday("Sunday",7):-!.

weekday(date(4,"November",2025),"Tuesday"):-!.
weekday(Date,DayName):- diffDays(Date,date(4,"November",2025),Diff), DayNumber is (abs(Diff) mod 7)+1, write(DayNumber), write(" "), weekday(DayName,DayNumber).

goal:-weekday(date(24,"December",2025) ,X), write(X), fail.

File: second_year/plf/exam_prep/course_material/Lecture_06/example_2.py
import sys

print(sys.executable)

import pyswip
from pyswip.easy import Functor, Variable, Atom
from pyswip.prolog import PrologError
from pyswip import Prolog, registerForeign, call, PL_new_atom, PL_new_functor, PL_cons_functor_v
# from datetime import datetime
import datetime

monthnumbers = {datetime.date(1,x,1).strftime('%B'):x for x in range(1,13)}

holidays_lst = [
    datetime.date(2025,1,1),
    datetime.date(2025,1,2),
    datetime.date(2025,1,6),
    datetime.date(2025,1,7),
    datetime.date(2025,1,24),
]

def holidays(date, handle):

    control = pyswip.core.PL_foreign_control(handle)

    index = None
    return_value = False

    if control == pyswip.core.PL_FIRST_CALL:
        index = 0

    if control == pyswip.core.PL_REDO:
        last_index = pyswip.core.PL_foreign_context(handle)
        index = last_index + 1

    if index >= 0 and index < len(holidays_lst):
        date.unify(formatDate(holidays_lst[index]))
        return_value = pyswip.core.PL_retry(index)

    return return_value


def formatDate(date):
    retval = Atom(f'date({date.day}, "{date.strftime("%B")}", {date.year})')
    return retval

def diffDays(date1, date2, diff):
    def date(day, month, year):
        try:
            month = month.decode("utf-8")
        except Exception:
            pass
        return datetime.date(year,monthnumbers[month],day)

    numVariables = sum([int(isinstance(x,Variable)) for x in [date1, date2, diff]])

    if numVariables == 0:
        date1 = eval(str(date1))
        date2 = eval(str(date2))
        return (date2-date1).days == diff
    elif numVariables == 1:
        if isinstance(diff,Variable):
            date1 = eval(str(date1))
            date2 = eval(str(date2))
            diff.value = (date2-date1).days
            return True
        elif isinstance(date1,Variable):
            date2 = eval(date2.value)
            new_date = date2-datetime.timedelta(days=diff)
            date1.unify(formatDate(new_date))
            return True
        elif isinstance(date2,Variable):
            date1 = eval(date1.value)
            new_date = date1+datetime.timedelta(days=diff)
            date2.unify(formatDate(new_date))
            return True
    else:
        raise PrologError("Not sufficienty instanciated")

def daysDiff(date1, date2):
    def date(day, month, year):
        try:
            month = month.decode("utf-8")
        except Exception:
            pass
        return datetime.date(year,monthnumbers[month],day)
    date1 = eval(str(date1))
    date2 = eval(str(date2))
    return (date2-date1).days


if __name__ == "__main__":

    registerForeign(diffDays)
    registerForeign(holidays, arity=1, flags=pyswip.core.PL_FA_NONDETERMINISTIC)

    Prolog.consult("example_2.pl")
    print('-'*80)

    for x in Prolog.query('successor(date(31,"December",2024),X)'):
        print(x)
    print('-'*80)

    # for x in Prolog.query('X, date(1,"December",2024),30)'):
    #     print(x)

    for x in Prolog.query('diffDays(date(31,"December",2024), date(1,"December",2024),X)'):
        print(x)
    print('-'*80)

    for x in Prolog.query('diffDays(X, date(1,"December",2024),-30)'):
        print(x)
    print('-'*80)

    for x in Prolog.query('diffDays(date(31,"December",2024), X,-30)'):
        print(x)
    print('-'*80)

    for x in Prolog.query('diffDays(date(31,"December",2024), date(1,"December",2024),-30)'):
        print(x)
    print('-'*80)

    for x in Prolog.query('diffDays(date(1,"December",2024), date(31,"December",2024),X)'):
        print(x)
    print('-'*80)

    for x in Prolog.query('diffDays(date(1,"December",2024), date(31,"December",2024),30)'):
        print(x)
    print('-'*80)

    for x in Prolog.query('diffDays(date(1,"December",2024), date(31,"December",2024),0)'):
        print(x)
    print('-'*80)

    for x in Prolog.query('daysFromDate(date(1,"December",2024), 30, date(31,"December",2024))'):
        print(x)
    print('-'*80)

    for x in Prolog.query('daysFromDate(X, 30, date(31,"December",2024))'):
        print(x)
    print('-'*80)

    for x in Prolog.query('daysFromDate(date(1,"December",2024), X, date(31,"December",2024))'):
        print(x)
    print('-'*80)

    for x in Prolog.query('daysFromDate(date(1,"December",2024), 30, X)'):
        print(x)
    print('-'*80)

    for x in Prolog.query('weekday(date(1,"December",1) ,X)'):
        print(x)
    print('-'*80)

    for x in Prolog.query('holidays(X)'):
        print(x)
    print('-'*80)

    for x in Prolog.query('holidays(X), diffDays(date(4,"November",2024), X, Y), weekday(X,Z)'):
        print(x)
    print('-'*80)